raven/src/main/java/com/getsentry/raven/connection/LockdownManager.java

```diff
@@ -0,0 +1,110 @@
+package com.getsentry.raven.connection;
+
+import com.getsentry.raven.time.Clock;
+import com.getsentry.raven.time.SystemClock;
+
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Abstracts the connection lockdown logic (and state) to a single place so that
+ * it's easier to understand.
+ */
+public class LockdownManager {
+    /**
+     * Default maximum duration for a lockdown.
+     */
+    public static final long DEFAULT_MAX_LOCKDOWN_TIME = TimeUnit.MINUTES.toMillis(5);
+    /**
+     * Default base duration for a lockdown.
+     */
+    public static final long DEFAULT_BASE_LOCKDOWN_TIME = TimeUnit.SECONDS.toMillis(1);
+    /**
+     * Maximum duration for a lockdown, in milliseconds.
+     */
+    private long maxLockdownTime = DEFAULT_MAX_LOCKDOWN_TIME;
+    /**
+     * Base duration for a lockdown, in milliseconds.
+     * <p>
+     * On each attempt the time is doubled until it reaches {@link #maxLockdownTime}.
+     */
+    private long baseLockdownTime = DEFAULT_BASE_LOCKDOWN_TIME;
+    /**
+     * Number of milliseconds after lockdownStartTime to lockdown for, or 0 if not currently locked down.
+     */
+    private long lockdownTime = 0;
+    /**
+     * Timestamp of when the current lockdown started, or null if not currently locked down.
+     */
+    private Date lockdownStartTime = null;
+    /**
+     * Clock instance used for time, injectable for testing.
+     */
+    private final Clock clock;
+
+    /**
+     * Construct a LockdownManager using the default system clock.
+     */
+    public LockdownManager() {
+        this(new SystemClock());
+    }
+
+    /**
+     * Construct a LockdownManager using the provided clock.
+     *
+     * @param clock Clock object to use for lockdown logic
+     */
+    public LockdownManager(Clock clock) {
+        this.clock = clock;
+    }
+
+    /**
+     * Returns true if the system is in a lockdown.
+     *
+     * @return true if the system is in a lockdown, otherwise false
+     */
+    public synchronized boolean isLockedDown() {
+        return lockdownStartTime != null && (clock.millis() - lockdownStartTime.getTime()) < lockdownTime;
+    }
+
+    /**
+     * Reset the lockdown state, disabling lockdown and setting the backoff time to zero.
+     */
+    public synchronized void resetLockdown() {
```

Comment:
Might be good for `resetLockdownState` for consistency?

Also since this is hanging off `LockdownManager`, maybe `resetState` would be sufficiently clear?